Fix: avoid error on empty search pattern (#1255)

pippotadde opened the pull request:
Fixes kiwix/kiwix-tools#737.
The server no longer throws a "no-query" error (400) when the pattern is empty, returning 200 OK instead.

Updates:
- I removed the check in "InternalServer::getSearchInfo".
- Updated "test/server.cpp" by removing expectations.
- Added "TEST_F(ServerTest, EmptyPatternSearchDoesNotError)".

veloman-yunkan reviewed (commented):
1. The ticket # mentioned in the commit is from a different repository, so the commit points to the wrong issue.

2. I think that the more important part of kiwix/kiwix-tools#737 is to prevent the search from happening when ENTER is pressed while the searchbox is empty. To this end you should modify the `performSearch()` function in `static/skin/viewer.js`. I will review the C++ changes when that part is ready too.

veloman-yunkan commented on src/server/internalServer.cpp:
```diff
@@ -334,26 +336,24 @@ SearchInfo InternalServer::getSearchInfo(const RequestContext& request) const
   if ( getLanguages(*mp_library, bookIds.second).size() != 1 ) {
     throw Error(nonParameterizedMessage("confusion-of-tongues"));
   }
-
```
Please undo this change

pippotadde commented:
Finally, everything should be OK now. I added and tested the pattern=yellow%20submarine test case, and I understood that URL-decoding in getZimQuery() was unnecessary, so I reverted that change and kept only trimming. Yesterday I was confused because of the existing pattern=j%C3%A0zz test case.

Updates:
- Split the work into the requested 5 commits
- Added the pattern=yellow%20submarine test case
- Restored getZimQuery() (no URL-decode, trim only)

veloman-yunkan commented on test/server.cpp:
```diff
@@ -77,7 +77,7 @@ const ResourceCollection resources200Compressible{
   { STATIC_CONTENT,  "/ROOT%23%3F/skin/taskbar.css?cacheid=42e90cb9" },
   { DYNAMIC_CONTENT, "/ROOT%23%3F/skin/viewer.js" },
-  { STATIC_CONTENT,  "/ROOT%23%3F/skin/viewer.js?cacheid=00e0fdf3" },
+  { STATIC_CONTENT,  "/ROOT%23%3F/skin/viewer.js?cacheid=8c3370b7" },
```
Regarding the commit message - it reads "Frontend: prevent empty search queries (update viewer.js cacheid test)". There is no need to include the note on cacheids in the message title/summary. If you want to preserve it, you can move it to the body of the message where more elaborate description and  additional details should belong.

veloman-yunkan commented on test/server.cpp:
```diff
@@ -456,6 +455,16 @@ TEST_F(ServerTest, 400)
+  // Whitespace-only pattern (should be trimmed)
+  r = zfs1_->GET("/ROOT%23%3F/search?content=zimfile&pattern=%20%20");
+  EXPECT_EQ(200, r->status);
+}
```
Why is this change in this commit? What does it have to do with the modification in the frontend? Would this new unit-test pass with code state corresponding to this commit?

veloman-yunkan commented on test/server.cpp:
```diff
@@ -1052,17 +1061,6 @@ TEST_F(ServerTest, Http400HtmlError)
-    </p>
-    <p>
-      No query provided.
-    </p>
```
Why have this and other test points been removed in this commit?

These changes should be moved to the last commit (the update in the backend)!

veloman-yunkan commented on test/server_search.cpp:
```diff
@@ -935,6 +953,36 @@ struct TestData
+      /* results */          YELLOW_SEARCH_RESULTS,
+      /* pagination */       {}
+    },
+
```
This change doesn't agree with the summary of the commit message. What does it have to do with data deduplication in existing unit-tests?

Would the tests pass if the code is compiled with the source tree checked out at this commit?

veloman-yunkan replied:
This change should be preserved. It just has to belong to the last commit.

veloman-yunkan commented on src/server/internalServer.cpp:
```diff
@@ -366,7 +362,8 @@ zim::Query SearchInfo::getZimQuery(bool verbose) const {
   }
-  query.setQuery(pattern);
+  const std::string finalPattern = kiwix::trim(pattern);
+  query.setQuery(finalPattern);
```
An extra variable is not justified. Call `kiwix::trim()` inline in the expression.

pippotadde commented:
History with the requested 5-commit structure:

1) Frontend: guard empty search input in viewer.js (+ necessary cacheid test updates).
2) Tests (refactor): deduplicate "yellow" search results (no behavior change).
3) Tests: add "yellow submarine" search case (using YELLOW_SEARCH_RESULTS[0]).
4) Tools: add kiwix::trim() + unit tests (including internal-whitespace case "abc 123").
5) Server: handle empty/whitespace patterns as valid input:
   - uses inline kiwix::trim() (no finalPattern)
   - returns 200 OK with empty results (instead of 400)
   - includes/moves the backend tests for empty/whitespace patterns
   - avoided unrelated formatting changes in internalServer.cpp

I verified that each commit builds and passes tests independently.

veloman-yunkan commented on test/server_search.cpp:
```diff
@@ -689,6 +689,24 @@ bool isSubSnippet(std::string subSnippet, const std::string& superSnippet)
+    /*wordCount*/  "204"
+  )
+};
+
```
This is the only change in this commit. `YELLOW_SEARCH_RESULTS` remains unused as of this commit (thus there is no deduplication, but actually additional duplication of search results data).

veloman-yunkan commented on test/server_search.cpp:
```diff
@@ -1045,29 +1045,13 @@ TEST(ServerSearchTest, searchResults)
     {
-      /* query */          "pattern=yellow&books.id=" RAYCHARLESZIMID,
-      /* start */            0,
+      /* query */          "pattern=yellow%20submarine&books.id=" RAYCHARLESZIMID,
```
The "yellow submarine" test point should have been added instead of replacing the "yellow" test point.

veloman-yunkan commented on src/server/internalServer.cpp:
```diff
@@ -986,6 +985,13 @@ std::unique_ptr<Response> InternalServer::handle_search_request(const RequestCon
+             q = zim::Query("__pippo_empty__");
+        } else {
+             q = searchInfo.getZimQuery(m_verbose.load());
+        }
```
What's this?

veloman-yunkan reviewed (changes requested):
1. A merge commit appeared as the first commit of this PR. Please get rid of it.
2. In the last commit, for a reason that I don't understand, the approach of fixing the handling of empty search patterns in the backend has changed and  `kiwix::trim()` remains unused.

veloman-yunkan commented:
Great!

Now let's go ahead with the next two commits (deduplication of search results for the "yellow" query and a new test point for "yellow submarine" query) as another PR.

pippotadde commented:
Thanks  :)

I created the separate PR #1259 for the two “yellow” test commits (deduplication + “yellow submarine” test).

veloman-yunkan commented:
Now please rebase this PR on recent main branch but preserve the remaining two commits in it:

1. Introduce `kiwix::trim()` with its unit-test
2. Fix the handling of empty search pattern in the backend via `kiwix::trim()` (like it was done in the last but one version of the PR) and update the server search unit tests (remove testpoints invalidated by the change in the server, add new testpoints for empty and/or whitespace-only search pattern).

pippotadde commented:
Hi, I rebased on main.
- Restored kiwix::trim()
- Fixed backend logic and tests as requested
- Verified locally

Let me know if there is anything else."

veloman-yunkan reviewed (changes requested):
Please also restore [these changes that you had in an earlier version of the PR](https://github.com/kiwix/libkiwix/pull/1255#discussion_r2642643603) - while doing so, please move the test point for " yellow " (with spaces) search pattern to sit amongst its other siblings (in fact you can just modify one of the duplicate test points).

veloman-yunkan commented:
Regarding the commit message of the second commit - as you can see it is too long and is wrapped by GitHub in an ugly way. Please follow the [guidelines for commit messages](https://stackoverflow.com/questions/2290016/git-commit-messages-50-72-formatting).

pippotadde commented:
Apologies for the confusion with the previous commit

Updates:
- Restored test cases
- Verified no whitespace-only changes
- Add /search?pattern case

veloman-yunkan reviewed (changes requested):
Almost there. Please fix the second commit as requested (you can use `git commit -a --amend`).

pippotadde commented:
Rebased, fixed tests and whitespace.

veloman-yunkan reviewed (changes requested):
I guess you are having problems with `git rebase`. In a sense `git` is a more important tool for software engineers than the compilers, interpreters, debuggers and other instruments directly related to their programming language of choice. Please dedicate some time to better learn `git` in general and `git rebase` in particular.
